Make neovim building even with libuv 1.18.0

mcepl opened the pull request:
(found for example on openSUSE/Leap 15)

I am not sure you care much about older systems, but this seems like a minimal pain for neovim building even on openSUSE/Leap 15 with libuv 1.18.0. More, I would like this to be here in case anybody needs it.

mcepl committed:
Make neovim building even with libuv 1.18.0

(found for example on openSUSE/Leap 15.*)

teto commented:
https://github.com/libuv/libuv/tree/v1.18.0 dates back from December 2017. Not super fan of this kind of patch: one can install a more recent version via flatpaks or nix (www.nixos.org). I'll let other maintainers chime in.

jamessan commented on src/nvim/event/stream.c:
```diff
@@ -16,6 +16,11 @@
 #endif
 
+// For compatbility with libuv < 1.19.0 (tested on 1.18.0)
+#ifndef uv_stream_get_write_queue_size
```
`uv_stream_get_write_queue_size` isn't a define, it's a function. It seems like this would _always_ be shadowing the actual function name.

A build-time check in CMakeLists.txt which sets our own define would be the proper way to handle this.

mcepl replied:
Could you please help this stupid Pythonista, who never even opened any `CMakeLists.txt` in his `$EDITOR`?

teto replied:
you would have to downgrade `find_package(LibUV 1.28.0 REQUIRED)` . 
Then you can use LibUV_VERSION_MINOR https://gitlab.kitware.com/cmake/cmake/commit/e56aa462976f80762712519a4cf653b8c45bf3db to conditionally add your define. 
(I am no cmake expert either so take it with a grain of salt).

jamessan replied:
> you would have to downgrade `find_package(LibUV 1.28.0 REQUIRED)` .

Hmm, not sure how @mcepl was able to build with these changes if we're requiring a libuv version >= 1.28 and using APIs that were introduced in 1.28.

jamessan replied:
> Hmm, not sure how @mcepl was able to build with these changes if we're requiring a libuv version >= 1.28 and using APIs that were introduced in 1.28.

To answer my own questions, the 1.28.x API use was reverted in #10488 but not the version requirement in CMakeLists.txt.  Also, our `find_package()` handling for libuv is apparently broken since it didn't fail the build when seeing that libuv 1.18 was installed but 1.28 was requested.

Prior to the use of `uv_gettimeofday()`, we had a comment suggesting our minimum libuv version was 1.12.

I'll work on cleaning up the CMake issues.

mcepl replied:
Thank you.

jamessan commented:
> Not super fan of this kind of patch: one can install a more recent version via flatpaks or nix ([www.nixos.org](http://www.nixos.org)). I'll let other maintainers chime in.

The stable PPA still builds against libuv versions older than 1.18, so I would also find it useful to avoid having to backport libuv.

mcepl commented:
Just that it may be a good idea to attach here [complete log of the build of 0.4.3 on openSUSE Leap 15.2 (libuv1-1.18.0-lp152.3.3)](https://github.com/neovim/neovim/files/4482347/_log.txt)
And this is [build on the same platform with the current master (commit 05fd64777)](https://github.com/neovim/neovim/files/4482740/_log.txt)

mcepl committed:
Test on actual libuv version number, not on existence of symbol.

uv_stream_get_write_queue_size isn't a define, it's a
function. It seems like this would always be shadowing the
actual function name.

jamessan reviewed (approved):
I'll handle updating the CMake version checks in a separate PR.  It looks like there's more work that needs to be done either in LibFindMacros.cmake or how we're using it.
